File: ktuan/ktuan.cpp
#include <stdio.h>

#define C 1000000000

using namespace std;

long long p[100010];
long a[1000], sign[1000];
long n;

void init() {
    long i, j = 1;
    for (i=1; ; i++) {
        if (i%2 == 1) {
            a[i] = j*(3*j-1) / 2;
            j = -j;
        } 
        else {
            a[i] = j*(3*j-1) / 2;
               j = (-j) + 1;        
        }
        if (a[i] > n) break;
    }    
    
    for (j=1; j<=i; j++) {
        if (j%4 == 1 || j%4 == 2) sign[j] = 1;
        else sign[j] = -1;
    }
}

void process() {
    p[0] = 1;
    //
    for (long i=1; i<= n; i++) {
        long j = 1;
        p[i] = 0;
        while (i-a[j] >= 0) {
            p[i] += p[i-a[j]]*sign[j];
            j++;            
        }
        p[i] = p[i]%C;
        while (p[i] < 0) p[i] += C;
    }
}

main() {
    scanf("%ld", &n);
    init();
    process();
    
    p[0] = 0;
    printf("%lld\n", p[n]);
    
    return 0;
}

File: ktuan.pas
uses math;
const   MAX_N = 100000;
        C = 1000000000;
var     fi,fo: text;
        a,sign: array[1..1000] of longint;
        p: array[0..MAX_N] of int64;
        n,time: longint;

procedure init;
var     i,j: longint;
begin
        j:= 1;
        for i:= 1 to n do begin
                if (i mod 2 = 1) then begin
                        a[i]:= j*(3*j-1) div 2;
                        j:= -j;
                end
                else begin
                        a[i]:= j*(3*j-1) div 2;
                        j:= (-j) + 1;
                end;
                if (a[i] > n) then break;
        end;
        //
        for j:= 1 to i do
         if (j mod 4 = 1) or (j mod 4 = 2) then sign[j]:= 1
         else sign[j]:= -1;
end;

procedure process;
var     i,j: longint;
begin
        p[0]:= 1;
        for i:= 1 to n do begin
                j:= 1;
                p[i]:= 0;
                while (i-a[j] >= 0) do begin
                        p[i]:= p[i]+p[i-a[j]]*sign[j];
                        inc(j);
                end;
                p[i]:= p[i] mod C;
                while (p[i] < 0) do p[i]:= p[i]+C;
        end;
end;

begin
        assign(fi,''); reset(fi);
        assign(fo,''); rewrite(fo);
        //
        readln(fi,n);
        init;
        process;
        p[0]:= 0;
        writeln(fo,p[n]);
        //
        close(fi);
        close(fo);
end.
